### DISCUSSION TITLE
Typed arrays

### DISCUSSION THREAD

**Discussion by Denneisk** [wiremod/wire#2840] [Suggestions]
Have arrays that can only be filled with a certain type. This allows for a performance improvement by not needing to type check on runtime. This also can help solve #2785 by implementing an alternative to vector2/4, ***however***, `array(123, 456)[number]` is a lot harder to write than `vec2(123, 456)`. Ideally, and with a future syntax, the compiler could coerce `[123, 456]` into a typed array when used as a literal in a function call.

Syntax?
```typescript
Array = array(...)[type]
Array = {...}[type]
Array = type[...] // Hmmm
```

The type system might need some extensions (generics, or just heavyhanded inheritance) to handle this, however.

**Comment by vurvdev** [comment]
The possibility of this is the only reason why arrays and tables aren't already using [1, 2, 3] or {1 = 2} syntax.

Lambdas too, but I think their usefulness outweighs the potential future typed version that would take another year or however long to actually implement.

**Reply by Denneisk** [reply]
Table syntax using `{}` might complicate the addition of JS objects if that were to ever be explored.

(Yes I happened to think about it again and wanted to say something for posterity)

**Reply by vurvdev** [reply]
Well yeah the typed arrays would be only [T...], I just wanted to show the syntax for typed tables as well.

Arrays shouldn't really have the key value constructor in the first place since they are meant to be sequential, but can't undo that
